File: h10/h10.rb
require 'date'
require_relative 'collection'

Pair = Struct.new(:fst,:snd)



def summingPairs(xs, sum)
    splitList =(xs.each_slice( (xs.size/2.0).round ).to_a)
    
    t1 = Thread.new {Thread.current[:output] = concurrentSummingPairHelper(splitList[0], splitList[0], sum)}
    t2 = Thread.new {Thread.current[:output] = concurrentSummingPairHelper(splitList[1], splitList[1], sum)}
    t3 = Thread.new {Thread.current[:output] = concurrentSummingPairHelperGreedy(splitList[0], splitList[1], sum)}
    t1.join
    t2.join
    t3.join
    
    return ( t1[:output]+t2[:output]+ t3[:output])
end

def concurrentSummingPairHelperGreedy(first_half, second_half, sum)
    the_pairs = []
    len = first_half.length
    len1 = second_half.length

    for i in 0..(len-1)
        for j in (0)..(len1-1)
        if first_half[i] + second_half[j] <= sum
            the_pairs.push(Pair.new(first_half[i],second_half[j]))
        end
        end
    end
    return the_pairs

end


def concurrentSummingPairHelper(first_half, second_half, sum)
    the_pairs = []
    len = first_half.length
    len1 = second_half.length

    for i in 0..(len-1)
        for j in (i+1)..(len1-1)
        if first_half[i] + second_half[j] <= sum
            the_pairs.push(Pair.new(first_half[i],second_half[j]))
        end
        end
    end
    return the_pairs

end



puts "Starting at:   #{DateTime.now.sec} seconds, #{DateTime.now.strftime("%9N")} nanoseconds"
puts summingPairs(INPUT,2020)
puts "Ending at:     #{DateTime.now.sec} seconds, #{DateTime.now.strftime("%9N")} nanoseconds"

File: h10/h10.clj
(load-file "./collection.clj")

(defn summingPairs [xs sum]
    (letfn [(summingPairsHelper [xs the_pairs]
            ;; If `xs` is empty, we're done.
        (if (empty? xs) the_pairs
            ;; Otherwise, decompose `xs` into the `fst` element
            ;; and the `rest`.
            (let [[fst & rest] xs]
                ;; We use the `recur` form to make the recursive call.
                ;; This ensures tail call optimisation
                (recur
                    rest
                    ;; Concatenate `the_pairs` we have so far with the sequence
                    ;; of every `[fst snd]` where `snd` is in `rest` with
                    ;; `fst + snd <= sum`. The `doall` outside the `concat`
                    ;; forces it to be calculated immediately; without this,
                    ;; we get a (lazy) buildup of `concat`'s which may
                    ;; cause a stack overflow when looking at the result.
                    (doall 
                        (concat the_pairs
                            (for [snd rest ;; For each `snd` in `rest`...
                                :when (<= (+ fst snd) sum)]
                            ;;... put `[fst snd]` into this sequence.
                            [fst snd])))))))]
    (summingPairsHelper xs [])))

(defn concurrentSummingPair [xs sum]

)



(println (str
            "Starting at:   "
            (.getSecond (java.time.LocalDateTime/now))
            " seconds, "
            (.getNano (java.time.LocalDateTime/now))
            " nanoseconds"))
(println (summingPairs input 2020))
(println (str
            "Ending at:     "
            (.getSecond (java.time.LocalDateTime/now))
            " seconds, "
            (.getNano (java.time.LocalDateTime/now))
            " nanoseconds"))